File: src/verification/verifier.ts
import { ethers } from 'ethers';
import {
  Challenge,
  ChallengeResponse,
  VerificationResult,
  NodeRegistration,
  HeartbeatRecord,
  LATENCY_THRESHOLDS,
} from '../types';
import { RpcClient, createRpcClient } from './rpc-client';
import { challengeGenerator } from './challenges';

export interface VerificationConfig {
  trustedRpcEndpoint: string;
}

export class NodeVerifier {
  private trustedRpc: RpcClient;
  private pendingChallenges: Map<string, { challenge: Challenge; expectedAnswer: string }> = new Map();

  constructor(config: VerificationConfig) {
    // We use our own trusted node to get the "correct" answers
    this.trustedRpc = createRpcClient(config.trustedRpcEndpoint);
  }

  // Create a challenge for a node to solve
  // We query our trusted node first so we know what the right answer is
  async createChallenge(node: NodeRegistration): Promise<Challenge> {
    const challenge = challengeGenerator.generateChallenge(node.id, node.nodeType);

    // Get the answer from our trusted node
    const expectedResponse = await this.trustedRpc.executeChallenge(challenge);

    if (!expectedResponse.success || !expectedResponse.data) {
      throw new Error(`Failed to get expected answer: ${expectedResponse.error}`);
    }

    // Store the challenge with its answer - we'll check against this later
    this.pendingChallenges.set(challenge.id, {
      challenge,
      expectedAnswer: expectedResponse.data,
    });

    // Send back the challenge WITHOUT the answer obviously
    return challenge;
  }

  // Check if a submitted answer is correct
  async verifyResponse(response: ChallengeResponse): Promise<VerificationResult> {
    const pending = this.pendingChallenges.get(response.challengeId);

    // Either we never issued this challenge or it already expired
    if (!pending) {
      return {
        challengeId: response.challengeId,
        nodeId: response.nodeId,
        passed: false,
        responseTimeMs: response.responseTimeMs,
        failureReason: 'Challenge not found or expired',
        timestamp: Date.now(),
      };
    }

    const { challenge, expectedAnswer } = pending;

    // Too slow - challenges expire after 1 minute
    if (Date.now() > challenge.expiresAt) {
      this.pendingChallenges.delete(response.challengeId);
      return {
        challengeId: response.challengeId,
        nodeId: response.nodeId,
        passed: false,
        responseTimeMs: response.responseTimeMs,
        failureReason: 'Challenge expired',
        timestamp: Date.now(),
      };
    }

    // Does their answer match ours?
    const answerMatches = this.compareAnswers(response.answer, expectedAnswer, challenge.type);

    if (!answerMatches) {
      this.pendingChallenges.delete(response.challengeId);
      return {
        challengeId: response.challengeId,
        nodeId: response.nodeId,
        passed: false,
        responseTimeMs: response.responseTimeMs,
        failureReason: 'Incorrect answer',
        timestamp: Date.now(),
      };
    }

    // Check if response time looks suspicious (might be using a public RPC)
    const latencyCheck = this.checkLatency(response.responseTimeMs);
    if (!latencyCheck.valid) {
      this.pendingChallenges.delete(response.challengeId);
      return {
        challengeId: response.challengeId,
        nodeId: response.nodeId,
        passed: false,
        responseTimeMs: response.responseTimeMs,
        failureReason: latencyCheck.reason,
        timestamp: Date.now(),
      };
    }

    // They passed!
    this.pendingChallenges.delete(response.challengeId);

    return {
      challengeId: response.challengeId,
      nodeId: response.nodeId,
      passed: true,
      responseTimeMs: response.responseTimeMs,
      timestamp: Date.now(),
    };
  }

  // Compare answers - different challenge types need different comparison logic
  private compareAnswers(submitted: string, expected: string, challengeType: string): boolean {
    const normalizedSubmitted = submitted.toLowerCase().trim();
    const normalizedExpected = expected.toLowerCase().trim();

    // Block hashes should match exactly
    if (challengeType === 'block-hash') {
      return normalizedSubmitted === normalizedExpected;
    }

    // Balances can have different formatting (leading zeros etc) so compare as numbers
    if (challengeType === 'state-balance') {
      try {
        return BigInt(normalizedSubmitted) === BigInt(normalizedExpected);
      } catch {
        return normalizedSubmitted === normalizedExpected;
      }
    }

    // JSON responses need to be parsed and compared properly
    if (challengeType === 'block-data' || challengeType === 'sync-status') {
      try {
        const submittedObj = JSON.parse(submitted);
        const expectedObj = JSON.parse(expected);
        return JSON.stringify(submittedObj) === JSON.stringify(expectedObj);
      } catch {
        return normalizedSubmitted === normalizedExpected;
      }
    }

    return normalizedSubmitted === normalizedExpected;
  }

  // Check if response time is suspiciously slow
  // Local nodes respond in <100ms, public RPCs take 300ms+
  private checkLatency(responseTimeMs: number): { valid: boolean; reason?: string } {
    if (responseTimeMs > LATENCY_THRESHOLDS.maxAllowed) {
      return { valid: false, reason: 'Response too slow - timeout' };
    }

    // Flag slow responses but don't fail them - could be network issues
    if (responseTimeMs > LATENCY_THRESHOLDS.suspiciousMin) {
      console.warn(`Suspicious latency: ${responseTimeMs}ms - might be proxying to public RPC`);
    }

    return { valid: true };
  }

  // Helper to verify wallet signatures
  async verifySignature(
    message: string,
    signature: string,
    expectedAddress: string
  ): Promise<boolean> {
    try {
      const recoveredAddress = ethers.verifyMessage(message, signature);
      return recoveredAddress.toLowerCase() === expectedAddress.toLowerCase();
    } catch {
      return false;
    }
  }

  // For nodes that expose their RPC, we can query them directly
  // This is more trusted than the local prover method
  async verifyExposedRpc(node: NodeRegistration): Promise<VerificationResult> {
    if (!node.rpcEndpoint) {
      return {
        challengeId: 'direct-' + Date.now(),
        nodeId: node.id,
        passed: false,
        responseTimeMs: 0,
        failureReason: 'No RPC endpoint configured',
        timestamp: Date.now(),
      };
    }

    const nodeRpc = createRpcClient(node.rpcEndpoint, node.authToken);

    // Generate a challenge
    const challenge = challengeGenerator.generateChallenge(node.id, node.nodeType);

    // Get the right answer from our trusted node
    const expectedResponse = await this.trustedRpc.executeChallenge(challenge);
    if (!expectedResponse.success || !expectedResponse.data) {
      return {
        challengeId: challenge.id,
        nodeId: node.id,
        passed: false,
        responseTimeMs: 0,
        failureReason: `Trusted node error: ${expectedResponse.error}`,
        timestamp: Date.now(),
      };
    }

    // Now ask their node the same question
    const userResponse = await nodeRpc.executeChallenge(challenge);

    if (!userResponse.success || !userResponse.data) {
      return {
        challengeId: challenge.id,
        nodeId: node.id,
        passed: false,
        responseTimeMs: userResponse.latencyMs,
        failureReason: userResponse.error || 'Failed to get response from node',
        timestamp: Date.now(),
      };
    }

    // Do the answers match?
    const answerMatches = this.compareAnswers(
      userResponse.data,
      expectedResponse.data,
      challenge.type
    );

    if (!answerMatches) {
      return {
        challengeId: challenge.id,
        nodeId: node.id,
        passed: false,
        responseTimeMs: userResponse.latencyMs,
        failureReason: 'Incorrect answer',
        timestamp: Date.now(),
      };
    }

    return {
      challengeId: challenge.id,
      nodeId: node.id,
      passed: true,
      responseTimeMs: userResponse.latencyMs,
      timestamp: Date.now(),
    };
  }

  // Quick check to see if a node is online and synced
  async checkHeartbeat(node: NodeRegistration): Promise<HeartbeatRecord | null> {
    if (!node.rpcEndpoint) {
      return null;
    }

    const nodeRpc = createRpcClient(node.rpcEndpoint, node.authToken);

    // Ask for block number, sync status, and peer count all at once
    const [blockResponse, syncResponse, peerResponse] = await Promise.all([
      nodeRpc.getBlockNumber(),
      nodeRpc.getSyncStatus(),
      nodeRpc.getPeerCount(),
    ]);

    if (!blockResponse.success) {
      return null;
    }

    return {
      nodeId: node.id,
      timestamp: Date.now(),
      blockNumber: blockResponse.data || 0,
      isSynced: syncResponse.success ? !syncResponse.data?.syncing : false,
      latencyMs: blockResponse.latencyMs,
      peersCount: peerResponse.success ? peerResponse.data || 0 : 0,
    };
  }

  // Remove old challenges that nobody answered
  cleanupExpiredChallenges(): number {
    const now = Date.now();
    let cleaned = 0;

    for (const [id, { challenge }] of this.pendingChallenges.entries()) {
      if (now > challenge.expiresAt) {
        this.pendingChallenges.delete(id);
        cleaned++;
      }
    }

    return cleaned;
  }
}

export function createVerifier(trustedRpcEndpoint: string): NodeVerifier {
  return new NodeVerifier({ trustedRpcEndpoint });
}

File: internal/verification/verifier.go
package verification

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"math/big"
	"strings"
	"sync"
	"time"

	"github.com/depinonbnb/depin/internal/challenge"
	"github.com/depinonbnb/depin/internal/rpc"
	"github.com/depinonbnb/depin/internal/types"
)

type pendingChallenge struct {
	Challenge      *types.Challenge
	ExpectedAnswer string
}

type Verifier struct {
	// trustedRPC is the quorum client added in Phase 3 per ADR-0009. It still
	// satisfies the same per-call surface as the old single *rpc.Client, but
	// returns RpcResponse{Error: rpc.QuorumNoMajority} when no majority is
	// reached. Callers below treat that as ABORT (skip + re-queue), NOT as
	// cheating.
	trustedRPC        *rpc.QuorumClient
	generator         *challenge.Generator
	pendingChallenges map[string]*pendingChallenge
	mu                sync.RWMutex
}

// NewVerifier constructs a Verifier whose trusted source is the quorum of
// the given RPC endpoints. Pass a single-element slice for ADR-0005
// behaviour; the QuorumClient logs a deprecation warning in that case.
func NewVerifier(trustedRPCEndpoints []string) *Verifier {
	return &Verifier{
		trustedRPC:        rpc.NewQuorum(trustedRPCEndpoints),
		generator:         challenge.NewGenerator(),
		pendingChallenges: make(map[string]*pendingChallenge),
	}
}

// TrustedRPC exposes the underlying QuorumClient so /ready can reach it for
// HealthCheck. Returning the live client (not a copy) keeps the readiness
// probe honest — anything the verifier sees, /ready sees too.
func (v *Verifier) TrustedRPC() *rpc.QuorumClient {
	return v.trustedRPC
}

// Create a challenge for a node
// We query our trusted node first so we know the right answer.
// Quorum-RPC abort path: if the trusted quorum can't agree (response.Error
// == rpc.QuorumNoMajority) we surface it as a normal error so the caller
// can decide what to do — per ADR-0009 the caller (scheduler / handler)
// must NOT punish the node, just skip and re-queue.
func (v *Verifier) CreateChallenge(node *types.NodeRegistration) (*types.Challenge, error) {
	ch := v.generator.GenerateChallenge(node.ID, node.NodeType)

	// Get the answer from our trusted node
	response := v.trustedRPC.ExecuteChallenge(ch)
	if !response.Success {
		if response.Error == rpc.QuorumNoMajority {
			slog.Warn("verification: trusted RPC quorum disagreement; aborting challenge create",
				"node_id", node.ID,
				"challenge_type", string(ch.ChallengeType),
			)
			return nil, fmt.Errorf("trusted RPC quorum disagreement (treated as abort, not cheating)")
		}
		return nil, fmt.Errorf("failed to get expected answer: %s", response.Error)
	}

	// Store the challenge with its answer
	v.mu.Lock()
	v.pendingChallenges[ch.ID] = &pendingChallenge{
		Challenge:      ch,
		ExpectedAnswer: response.Data,
	}
	v.mu.Unlock()

	return ch, nil
}

// Check if a submitted answer is correct
func (v *Verifier) VerifyResponse(response *types.ChallengeResponse) *types.VerificationResult {
	v.mu.RLock()
	pending, exists := v.pendingChallenges[response.ChallengeID]
	v.mu.RUnlock()

	now := time.Now().UnixMilli()

	// Challenge not found or already used
	if !exists {
		return &types.VerificationResult{
			ChallengeID:    response.ChallengeID,
			NodeID:         response.NodeID,
			Passed:         false,
			ResponseTimeMs: response.ResponseTimeMs,
			FailureReason:  "challenge not found or expired",
			Timestamp:      now,
		}
	}

	// Too slow - challenges expire after 1 minute
	if now > pending.Challenge.ExpiresAt {
		v.deleteChallenge(response.ChallengeID)
		return &types.VerificationResult{
			ChallengeID:    response.ChallengeID,
			NodeID:         response.NodeID,
			Passed:         false,
			ResponseTimeMs: response.ResponseTimeMs,
			FailureReason:  "challenge expired",
			Timestamp:      now,
		}
	}

	// Does their answer match ours?
	if !v.compareAnswers(response.Answer, pending.ExpectedAnswer, pending.Challenge.ChallengeType) {
		v.deleteChallenge(response.ChallengeID)
		return &types.VerificationResult{
			ChallengeID:    response.ChallengeID,
			NodeID:         response.NodeID,
			Passed:         false,
			ResponseTimeMs: response.ResponseTimeMs,
			FailureReason:  "incorrect answer",
			Timestamp:      now,
		}
	}

	// Check if response time looks suspicious
	if response.ResponseTimeMs > types.LatencyMaxAllowed {
		v.deleteChallenge(response.ChallengeID)
		return &types.VerificationResult{
			ChallengeID:    response.ChallengeID,
			NodeID:         response.NodeID,
			Passed:         false,
			ResponseTimeMs: response.ResponseTimeMs,
			FailureReason:  "response too slow",
			Suspicious:     true,
			SuspiciousNote: "Response took too long - possible proxy or offline node",
			Timestamp:      now,
		}
	}

	// They passed!
	v.deleteChallenge(response.ChallengeID)

	// Flag slow responses but still pass them (suspicious but not failed)
	suspicious := response.ResponseTimeMs > types.LatencySuspiciousMin
	suspiciousNote := ""
	if suspicious {
		suspiciousNote = fmt.Sprintf("High latency %dms - might be proxying to public RPC", response.ResponseTimeMs)
		slog.Warn("suspicious latency",
			"node_id", response.NodeID,
			"response_time_ms", response.ResponseTimeMs,
		)
	}

	return &types.VerificationResult{
		ChallengeID:    response.ChallengeID,
		NodeID:         response.NodeID,
		Passed:         true,
		ResponseTimeMs: response.ResponseTimeMs,
		Suspicious:     suspicious,
		SuspiciousNote: suspiciousNote,
		Timestamp:      now,
	}
}

func (v *Verifier) deleteChallenge(id string) {
	v.mu.Lock()
	delete(v.pendingChallenges, id)
	v.mu.Unlock()
}

// Compare answers - different challenge types need different comparison
func (v *Verifier) compareAnswers(submitted, expected string, challengeType types.ChallengeType) bool {
	submitted = strings.ToLower(strings.TrimSpace(submitted))
	expected = strings.ToLower(strings.TrimSpace(expected))

	switch challengeType {
	case types.BlockHash:
		// Block hashes should match exactly
		return submitted == expected

	case types.StateBalance:
		// Balances can have different formatting so compare as numbers
		subBig, ok1 := new(big.Int).SetString(strings.TrimPrefix(submitted, "0x"), 16)
		expBig, ok2 := new(big.Int).SetString(strings.TrimPrefix(expected, "0x"), 16)
		if ok1 && ok2 {
			return subBig.Cmp(expBig) == 0
		}
		return submitted == expected

	case types.BlockData, types.SyncStatus:
		// JSON responses need to be parsed and compared
		var subObj, expObj map[string]interface{}
		if json.Unmarshal([]byte(submitted), &subObj) == nil &&
			json.Unmarshal([]byte(expected), &expObj) == nil {
			subJSON, _ := json.Marshal(subObj)
			expJSON, _ := json.Marshal(expObj)
			return string(subJSON) == string(expJSON)
		}
		return submitted == expected

	default:
		return submitted == expected
	}
}

// For nodes that expose their RPC, we query them directly
func (v *Verifier) VerifyExposedRPC(node *types.NodeRegistration) *types.VerificationResult {
	now := time.Now().UnixMilli()

	if node.RPCEndpoint == "" {
		return &types.VerificationResult{
			ChallengeID:   fmt.Sprintf("direct-%d", now),
			NodeID:        node.ID,
			Passed:        false,
			FailureReason: "no RPC endpoint configured",
			Timestamp:     now,
		}
	}

	nodeRPC := rpc.NewClient(node.RPCEndpoint, node.AuthToken)

	// Generate a challenge
	ch := v.generator.GenerateChallenge(node.ID, node.NodeType)

	// Get the right answer from our trusted node
	expectedResponse := v.trustedRPC.ExecuteChallenge(ch)
	if !expectedResponse.Success {
		// ADR-0009 abort path: a "no quorum" answer means the trusted side
		// disagreed with itself. We MUST NOT bump the operator's warning
		// count — Suspicious=false guarantees the in-memory + sqlite stores
		// won't escalate. Caller still records the result so it shows up in
		// history; downstream stats can ignore it.
		if expectedResponse.Error == rpc.QuorumNoMajority {
			slog.Warn("verification: trusted RPC quorum disagreement; aborting verify",
				"node_id", node.ID,
				"challenge_type", string(ch.ChallengeType),
			)
			return &types.VerificationResult{
				ChallengeID:   ch.ID,
				NodeID:        node.ID,
				Passed:        false,
				FailureReason: "trusted RPC quorum disagreement (treated as abort, not cheating)",
				Suspicious:    false,
				Timestamp:     now,
			}
		}
		return &types.VerificationResult{
			ChallengeID:   ch.ID,
			NodeID:        node.ID,
			Passed:        false,
			FailureReason: fmt.Sprintf("trusted node error: %s", expectedResponse.Error),
			Timestamp:     now,
		}
	}

	// Now ask their node the same question
	userResponse := nodeRPC.ExecuteChallenge(ch)
	if !userResponse.Success {
		return &types.VerificationResult{
			ChallengeID:    ch.ID,
			NodeID:         node.ID,
			Passed:         false,
			ResponseTimeMs: userResponse.LatencyMs,
			FailureReason:  userResponse.Error,
			Timestamp:      now,
		}
	}

	// Do the answers match?
	if !v.compareAnswers(userResponse.Data, expectedResponse.Data, ch.ChallengeType) {
		return &types.VerificationResult{
			ChallengeID:    ch.ID,
			NodeID:         node.ID,
			Passed:         false,
			ResponseTimeMs: userResponse.LatencyMs,
			FailureReason:  "incorrect answer",
			Timestamp:      now,
		}
	}

	return &types.VerificationResult{
		ChallengeID:    ch.ID,
		NodeID:         node.ID,
		Passed:         true,
		ResponseTimeMs: userResponse.LatencyMs,
		Timestamp:      now,
	}
}

// Quick check to see if a node is online and synced
func (v *Verifier) CheckHeartbeat(node *types.NodeRegistration) *types.HeartbeatRecord {
	if node.RPCEndpoint == "" {
		return nil
	}

	nodeRPC := rpc.NewClient(node.RPCEndpoint, node.AuthToken)

	blockNum, latency, err := nodeRPC.GetBlockNumber()
	if err != nil {
		return nil
	}

	synced, _, _ := nodeRPC.GetSyncStatus()
	peerCount, _, _ := nodeRPC.GetPeerCount()

	return &types.HeartbeatRecord{
		NodeID:      node.ID,
		Timestamp:   time.Now().UnixMilli(),
		BlockNumber: blockNum,
		IsSynced:    synced,
		LatencyMs:   latency,
		PeersCount:  peerCount,
	}
}

// Remove old challenges that nobody answered
func (v *Verifier) CleanupExpiredChallenges() int {
	now := time.Now().UnixMilli()
	cleaned := 0

	v.mu.Lock()
	defer v.mu.Unlock()

	for id, pending := range v.pendingChallenges {
		if now > pending.Challenge.ExpiresAt {
			delete(v.pendingChallenges, id)
			cleaned++
		}
	}

	return cleaned
}
